src/Agent.Plugins/PipelineArtifact/FileContainerProvider.cs

```diff
@@ -0,0 +1,191 @@
+        {
+            var containerIdAndRoot = ParseContainerId(artifact.Resource.Data);
+            
+            var items = await containerClient.QueryContainerItemsAsync(
+                containerIdAndRoot.Item1,
+                projectId,
+                containerIdAndRoot.Item2
+                );
+
+            tracer.Info($"Start downloading FCS artifact- {artifact.Name}");
+            IEnumerable<Func<string, bool>> minimatcherFuncs = MinimatchHelper.GetMinimatchFuncs(minimatchPatterns, tracer);
+            if (minimatcherFuncs !=null && minimatcherFuncs.Count() !=0)
+            {
+                items = this.GetFilteredItems(items, minimatcherFuncs, containerIdAndRoot.Item2);
+            }
+
+            var folderItems = items.Where(i => i.ItemType == ContainerItemType.Folder);
+            Parallel.ForEach(folderItems, (folder) =>
+            {
+                var targetPath = ResolveTargetPath(rootPath, folder, containerIdAndRoot.Item2);
+                Directory.CreateDirectory(targetPath);
+            });
+
+            var fileItems = items.Where(i => i.ItemType == ContainerItemType.File);
+
+            var downloadBlock = NonSwallowingActionBlock.Create<FileContainerItem>(
+                async item =>
+                {
+                    var targetPath = ResolveTargetPath(rootPath, item, containerIdAndRoot.Item2);
+                    await AsyncHttpRetryHelper.InvokeVoidAsync(
+                       async () =>
+                       {
+                           using (var sourceStream = await this.DownloadFileFromContainerAsync(containerIdAndRoot, projectId, containerClient, item, cancellationToken))
+                           {
+                               tracer.Info($"Downloading: {targetPath}");
+                               using (var targetStream = new FileStream(targetPath, FileMode.Create))
+                               {
+                                   sourceStream.CopyTo(targetStream);
+                               }
+                           }
+                       },
+                        maxRetries: 5,
+                        cancellationToken: cancellationToken,
+                        tracer: tracer,
+                        continueOnCapturedContext: false,
+                        canRetryDelegate: exception => exception is IOException,
+                        context: null
+                        );
+                },
+                new ExecutionDataflowBlockOptions()
+                {
+                    BoundedCapacity = Environment.ProcessorCount * 8,
+                    MaxDegreeOfParallelism = 8,
+                    CancellationToken = cancellationToken,
+                });
+
+            foreach(var item in fileItems)
+            {
+                await downloadBlock.SendAsync(item);
```

Comment:
After calling SendAsync, you need wait on the block's completion. I think we have a helper method to do send all and wait.